examples/stats_models/regression.jl

```diff
@@ -0,0 +1,243 @@
+
+using SemanticModels
+using Random
+using LightGraphs
+using MetaGraphs
+using DataFrames
+using GraphPlot
+using DifferentialEquations
+using Plots
+using MultivariateStats
+using Statistics
+
+mutable struct DepVar{T}
+    value::T
+end
+
+mutable struct IndepVar{T}
+    value::T
+end
+
+# to allow transformations that feature numeric elements/vectors 
+mutable struct Other{T}
+    value::T
+end
+
+
+const Var = Union{DepVar, IndepVar}
+const RegComponents = Union{DepVar, IndepVar, Other}
+
+
+mutable struct OpElementTypes{a, b}
+    vec_a::a
+    vec_b::b
+end
+
+function var_type_check(a::Var, b::Other)
+    return isa(a, DepVar) ? DepVar : IndepVar
+end
+    
+function var_type_check(a::Other, b::Var)
+    return isa(b, DepVar) ? DepVar : IndepVar
+end
+
+function var_type_check(a::Var, b::Var)   
+    return isa(a, DepVar) || isa(b, DepVar) ? DepVar : IndepVar
+
+end
+
+function var_type_check(a::Other, b::Other)   
+    return Other
+end
+
+
+binary_ops = [+,-,*,/]
+
+# example loops to get element-level truth tables 
+dv_num = 10
+iv_num = 3
+other_num = 7
+
+
+for op in binary_ops
+    for a in (DepVar(dv_num), IndepVar(iv_num), Other(other_num))
+        for b in (DepVar(dv_num), IndepVar(iv_num), Other(other_num))
+            output_check = var_type_check(a,b)
+            result_type = output_check(op(a.value, b.value))
+            print("Op: $op, a: $a, b: $b, result_type: $result_type \n")
+        end
+    end
+end
+
+function transform_col_vector(op, vec_a::Array, vec_b::Array)
+    
+    if length(vec_a) == length(vec_b)
+
+        output_types = map(var_type_check, [x for x in vec_a], [x for x in vec_b]) 
+        numeric_results = broadcast(op, [x.value for x in vec_a], [x.value for x in vec_b])    
+        typed_results = [i(j) for (i,j) in zip(output_types, numeric_results)]
+            #output_types[i](numeric_results[i]) for i in range](1:length(vec_a))]
+        return typed_results
+    else
+        return "Error: dimensions mismatch"
+    end
+end
+
+function regress(vec_a::Array{IndepVar{Float64}, 2}, vec_b::Array{IndepVar{Float64}, 2}, orig_x::Array{IndepVar{Float64}, 2}, orig_y::Array{DepVar{Float64}, 2}, op)
```

Comment:
To make these more generic, you can use this construction over Float64:
```julia
julia> function foo(a::Vector{T}) where {T<:Number}
         return a .+ 1
         end
foo (generic function with 1 method)

julia> foo([1,2,3])
3-element Array{Int64,1}:
 2
 3
 4

julia> foo([1,2,3.0])
3-element Array{Float64,1}:
 2.0
 3.0
 4.0
```

Reply:
Actual I think this is what they call "triangular dispatch"

`Array{T}` and then `where {T <: DepVar}` and then you don't have the constraint that `DepVar{U} where U<:Number`

Reply:
Maybe this is a way to do it


Define a parametric struct
```julia
julia> struct Bar{T}
       data::T
       end

# define a method that expects Bar{T<:Number}
julia> function foo(b::Bar{T}) where T<:Number
       return b.data .+ 1
       end
foo (generic function with 2 methods)

# it works, this is the easy case
julia> foo(Bar(1))
2

# it doesn't work if T is not a subtype of number
julia> foo(Bar([1,2]))
ERROR: MethodError: no method matching foo(::Bar{Array{Int64,1}})
Closest candidates are:
  foo(::Array{T<:Number,1}) where T<:Number at REPL[4]:2
  foo(::Bar{T<:Number}) where T<:Number at REPL[10]:2
Stacktrace:
 [1] top-level scope at none:0

# define a case for Bar of Array containing numbers
julia> function foo(b::Bar{Array{T,1}}) where T<:Number
       return b.data .+ 1
       end
foo (generic function with 3 methods)

# this works so nested type constraints are possible.
julia> foo([1,2,3])
3-element Array{Int64,1}:
 2
 3
 4

# now test that the array can be on the outside and the parametric struct can be on the inside.
julia> function foo(b::Array{Bar{T},1}) where T<:Number
       return [x.data .+ 1 for x in b]
       end
foo (generic function with 4 methods)

# success
julia> foo(Bar.([1,2,3]))
3-element Array{Int64,1}:
 2
 3
 4

```

Reply:
So  a relatively straightforward fix seems to work unless there's something I've overlooked:

`regress(vec_a::Array{IndepVar{T}}, vec_b::Array{DepVar{T}}, orig_x::Array{IndepVar{T}}, orig_y::Array{DepVar{T}}, op) where {T<:Number}`

e.g., constrain for the RegComponent parameter type but don't include a constraint that the array dimension parameter <: Integer.